File: Algorithms/golang-sorting/sort/quicksort/quicksort.go
package quicksort

import "heartiger.com/algorithms/utility"


// Quicksort set a pivot from the array
// Put all elements <= pivot to the left
// And elements > pivot to the right of the pivot.
// Then sort left hand side and right hand side until left=right.
// Time Complexity O(nlogn)
func QuickSorter(nums []int) {
	quickSort(nums, 0, len(nums))
}
func quickSort(nums []int, left int, right int) {

	if left >= right {
		return
	}

	partitionIndex := partition(nums, left, right)
	quickSort(nums, left, partitionIndex)
	quickSort(nums, partitionIndex+1, right)
}
func partition(nums []int, left int, right int) int {
	pivot := nums[right-1]

	largerItemIndex := left
	for i := left; i < right-1; i++ {
		if nums[i] > pivot {
			continue
		}
		utility.Swap(nums, largerItemIndex, i)
		largerItemIndex++
	}
	utility.Swap(nums, largerItemIndex, right-1)
	return largerItemIndex
}

File: Algorithms/Sorting/src/sorters/QuickSort.java
package sorters;

import utility.Sortable;
import utility.Utility;

public class QuickSort implements Sortable {

    private int[] nums;

    public QuickSort() {
    }

    public QuickSort(int[] nums) {
        this.nums = nums;
    }


    @Override
    public int[] sort() {
        sortArray(0, nums.length - 1);
        return nums;
    }

    private void sortArray(int left, int right) {
        if (left >= right)
            return;

        int partitionIndex = partition(left, right);
        sortArray(left, partitionIndex - 1); // Exclude element at partitionIndex
        sortArray(partitionIndex + 1, right);
    }

    private int partition(int left, int right) {
        int pivot = nums[right];

        int largerItemIndex = left;
        for (int i = left; i < right; i++) {
            if (nums[i] <= pivot) {
                Utility.swap(nums, largerItemIndex, i);
                largerItemIndex++;
            }
        }
        Utility.swap(nums, largerItemIndex, right);
        return largerItemIndex;
    }

    @Override
    public void setArray(int[] nums) {
        this.nums = nums;
    }
}
